### ISSUE TITLE
AMD IOMMU initializing GPU failed at if(!dev_data->passthrough) when amdkfd loads GPU device

### ISSUE THREAD

**Issue by Yougmark** [RadeonOpenCompute/ROCK-Kernel-Driver#42]
Hi,

I've been trying to get Vega in Raven Ridge APU running (on HP new ENVY 360).  So far, I tracked down where it fails: kfd failed to initialize iommu because the device struct data indicates that passthrough is not supported.  Specifically, it fails at [line 2306, file drivers/iommu/amd_iommu.c](https://github.com/RadeonOpenCompute/ROCK-Kernel-Driver/blob/2691bd0b3c79b717951e7a0fdbe7e5cbac951bb5/drivers/iommu/amd_iommu.c#L2306).

Any suggestions on fixing this?

``dmesg`` snippets are attached, showing kfd failed on iommu.  Although werid, GPU is recognized as a discrete one.

```
[    1.096291] AMD IOMMUv2 driver by Joerg Roedel <jroedel@suse.de>
[    1.097325] Parsing CRAT table with 1 nodes
[    1.097327] Ignoring ACPI CRAT on non-APU system
[    1.097328] Virtual CRAT table created for CPU
[    1.097329] Parsing CRAT table with 1 nodes
[    1.097329] Creating topology SYSFS entries
[    1.097336] Topology: Add CPU node
[    1.097336] Finished initializing topology
[    1.099346] kfd kfd: Initialized module
[    1.099494] checking generic (e0000000 7f0000) vs hw (e0000000 10000000)
[    1.099495] fb: switching to amdgpudrmfb from VESA VGA
```

```
[    1.837632] kfd kfd: Allocated 3969056 bytes on gart
[    1.837693] Virtual CRAT table created for GPU
[    1.837693] Parsing CRAT table with 1 nodes
[    1.837696] Creating topology SYSFS entries
[    1.837742] Topology: Add dGPU node [0x15dd:0x1002]
[    1.837862] kfd kfd: Reserved 2 pages for cwsr.
[    1.837871] kfd kfd: failed to initialize iommu
[    1.837874] kfd kfd: Error resuming kfd
[    1.837886] Creating topology SYSFS entries
[    1.838132] kfd kfd: device 1002:15dd NOT added due to errors
[    1.838142] [drm] Initialized amdgpu 3.20.0 20150101 for 0000:04:00.0 on minor 0
```

In addition, here is my vector_copy output:
```
Initializing the hsa runtime succeeded.
Checking finalizer 1.0 extension support succeeded.
Generating function table for finalizer succeeded.
Getting a gpu agent failed.
```
I fixed a few problems in ROCT and ROCR here and there to make it reach this point of adding gpu agent. However, gpu_agents_ is a vector of size 0 because GPU is not initialized successfully.

Any help is appreciated! Thanks!

**Comment by whchung** [comment]
@jedwards-AMD may know better but to my understanding raven ridge support is not implemented in the released version of ROCm stack yet.

**Comment by Yougmark** [comment]
Thanks for the prompt reply. Yeah, I got that message too.  But still trying to get it running, probably I'm being naive.  :(
What is this passthrough indicating, though? Virtualization? Comments in the file say "device is identity mapped".
Btw, do you have a schedule to release full support for raven ridge, approximately?
Thanks!

**Comment by fxkamd** [comment]
I think there are a few different problems with this system. KFD ignores the CRAT table because it doesn't contain any information about the GPU. That's a problem in the system BIOS. As a result it wants to treat the GPU like a discrete GPU rather than an APU. But then it detects the Raven GPU and realizes it needs IOMMU anyway, and fails.

The IOMMU driver intialization problem is something I'm not too familiar with. I suspect there may be some BIOS setting that prevents this from working. Maybe IOMMU is disabled in your BIOS setup.

All that said, we're still working through some IOMMU driver issues on Raven. So even if you could get it to work, it wouldn't be stable and it's not supported in current ROCm releases yet.

**Comment by Yougmark** [comment]
Thanks for the reply. I guess I need to figure out how to enable IOMMU in BIOS, which looks like not straight forward with this HP laptop... And then look forward to the new ROCm release.

Can you provide an approximate time of the release time? Mid/Later 2018? Early 2019?

Thanks,

**Comment by Yougmark** [comment]
Setting up ACPI CRAT table looks like a pain, after looking around for a while.  Does anyone have any suggestions on this? My machine is HP envy x360 with raven ridge.

Thanks,

**Comment by fxkamd** [comment]
The CRAT table, and disabled IOMMU in the BIOS has been an issue on many Carrizo laptops too. I think it's unrealistic to expect that OEMs will release updated BIOS images with fixes for that, because these features aren't used at all on Windows, which is typically the only OS the OEMs officially support.

It should be possible to work around the problems in KFD. Below I'm stating what I think is possible, but I can't promise that AMD will be working on this. AMD is upstreaming the KFD driver into the mainline Linux kernel, which will simplify 3rd-party contributions.

If the CRAT table is missing, we should be able to generate one ourselves. We do that for discrete GPUs and CPUs. It shouldn't be terribly difficult to do the same for APUs.

If the IOMMUv2 cannot enabled, we should be able to treat the APU like a dGPU without its own memory. That would allow us to use the dGPU memory allocation and mapping code paths that don't depend on the IOMMUv2 to do virtual address translations. But it would affect the programming model and would probably require changes in the higher level components as well (OpenCL, HCC, HIP, etc.). So this would be much more involved.

**Comment by Yougmark** [comment]
Thanks for the reply!
```
$ ps -ef | grep iommu
root       181     2  0 19:20 ?        00:00:00 [amd_iommu_v2]
$ lsmod | grep iommu
amd_iommu_v2           20480  1 amdkfd
```

I think this verified that iommu_v2 is up. Also in the dmesg log, ``AMD IOMMUv2 driver by Joerg Roedel <jroedel@suse.de>`` shows it's loaded without further error. Please let me know if I'm wrong.

Regarding CRAT, currently, we are trying to see if we can contact HP to get the BIOS with valid CRAT for APU. In this [issue thread](https://github.com/RadeonOpenCompute/ROCm/issues/106), I found people successfully use one generation older APU (bristol bridge) on a one generation older HP envy x360 laptop (i.e., same product line but one generation older). It's surprising, but it means that HP is able to configure CRAT correctly in BIOS. I don't know why they didn't on this newer machine.

Regarding the working around in kfd, I'm not confident to make it work for APU, but I might give it a try.  Is your branch the right place to start: [fkxamd/drm-next-wip](https://github.com/RadeonOpenCompute/ROCK-Kernel-Driver/tree/fkxamd/drm-next-wip). I've been reading your emails of patches fixing problems for dGPU in amdkfd in the mailing list amd-gfx@lists.freedesktop.org.  What's the repository for this mailing list? Is it this repository?

Some sense of where we are heading. We are really interested in getting APU running for our research work.  Our research interests are real-time GPU, mainly for the autonomous system. The last dissertation written by an upper-class student in our group is centered around this [GPUSync](http://www.cs.unc.edu/~anderson/papers/rtss13c.pdf) paper. We are continuing our work to improve real-time predictability for GPU.  Open source stack provided by AMD looks promising. The heterogeneous platform is always interested for the embedded system although APU is not in an embedded system yet.

**Comment by fxkamd** [comment]
The amd_iommu_v2 driver being loaded doesn't mean that the IOMMU is present and working. This is loaded even on my dGPU test system with an Intel GPU. My log does say this though: "AMD IOMMUv2 functionality not available on this system".

The failure in the headline of this issue indicates a problem either with the IOMMUv2 hardware or driver. I'm not terribly familiar with either. Which kernel version are you using? Maybe try a newer one. There is still work going on to get the IOMMUv2 driver work correctly on Raven.

drm-next-wip is where I'm doing the upstreaming. It will be a while longer until that branch supports GFXv9 (Vega10, Raven) and gets accepted upstream. Currently I'm aiming for Linux 4.17 for GFXv8 and probably 4.18 for GFXv9.

Until then you'll have to stick with the ROCm release branches. ROCm 1.7 is the latest release.

**Comment by Yougmark** [comment]
My log doesn't have the line you quoted.  Here are the lines containing IOMMU from the dmesg log.
```
$ dmesg | grep IOMMU
[    4.483290] AMD-Vi: Unable to write to IOMMU perf counter.
[    4.485565] AMD-Vi: Found IOMMU at 0000:00:00.2 cap 0x40
[    5.263025] AMD IOMMUv2 driver by Joerg Roedel <jroedel@suse.de>
[    9.890065] vboxpci: IOMMU found
```

I'm using the kernel from the master branch of this repository, which is the kernel for ROCm v1.7.

Later this week, I'll dig deeper to get a better understanding of the problem hopefully.

One last quick question. My HP laptop is Ryzen 5 2500U, Radeon Vega 8 as it's marketed.  And I never found the code number in Linux, not in dmesg log and lshw output. But the released firmware and the gfx code under driver/drm/amd/ are all using code Vega 10.  Is this just a code name mismatch or is Vega 8 really not supported?  Would this matter, basically?

Thanks,
Ming

**Comment by fxkamd** [comment]
I suspect Vega8 is what they ended up calling the integrated GPU in the Raven (codename) APU (now marketed as Ryzen 5). We tend not to put marketing names into the driver, because they tend to change or be decided by Marketing a short time before the release when the driver is already mostly finished.

Vega10 was in fact a code name, which is why you find it in the driver. I haven't heard the name Vega 8 before. I guess Vega crossed over from being a code name to being a marketing name, confusing every one. You'll find the code name Raven in the driver, to refer to the integrated GPU in your Ryzen APU.

**Comment by nealeu** [comment]
Thanks for pushing on with this.  I just tried the latest nightly of Ubuntu 18.04 LTS from a USB stick on a borrowed Envy 360 Ryzen 2500U, and it hung with the following on startup:

    AMD-Vi: Unable to write to IOMMU perf counter.
    amd_gpio AMDI0030:00: Failed to get gpio IRQ: -22

So does that mean that we're realistically looking at 4.17 (or higher) in 18.04.1 for good out of the box support as far as Ubuntu goes, or should we expect 4.15 to get patches to at least boot?

**Comment by Yougmark** [comment]
@nealeu I met the same problem. Sometimes it does not hang.  Felix told me Raven IOMMU isn't well supported and the latest fixes should go into 4.16.

**Comment by Johnreidsilver** [comment]
From other threads it's clear we'll have to wait for Raven Ridge support.

Ubuntu 18.04 on Raven Ridge R3 2200g, IOMMU enabled in BIOS

```
uname -a
Linux r3 4.16.0-041600rc4-generic #201803041930 SMP Mon Mar 5 00:32:34 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

```
 dmesg | grep IOMMU
[    0.694963] AMD-Vi: Unable to write to IOMMU perf counter.
[    0.698105] AMD-Vi: Found IOMMU at 0000:00:00.2 cap 0x40
[    1.065587] AMD IOMMUv2 driver by Joerg Roedel <jroedel@suse.de>
```

**Comment by Yougmark** [comment]
@Johnreidsilver what error did you met if you tried some GPU program?  It seems your amdkfd driver is initialized okay.  I'm not sure "Unable to write to IOMMU perf counter" will prevents you from running GPU programs.  My situation was not being able to setup kfd virtual device.  But you should be able to see them setup correctly?  Btw, which machine you are using with the correct CRAT?

@fxkamd hi Felix, how is the process of IOMMU problem fix?  Can we grab the up-to-date code somewhere?  This repo and your WIP branch in this repo haven't been updated for a while.

Thanks,
Ming

**Comment by fxkamd** [comment]
@Yougmark : My drm-next-wip branch was updated about 10 days ago. It's currently based on 4.16-rc1. It looks like the fix I was referring to was merged: https://github.com/RadeonOpenCompute/ROCK-Kernel-Driver/commit/ff18c4e598de13af6503d6adb66f3ad768b6a53e#diff-6f21dc60b8c807045e0dcac61ba57bef

**Comment by Johnreidsilver** [comment]
This Raven Ridge APU (R3 2200G, motherboard ASUS A320M-K Bios v.3803), still missing some basic support like OpenCL. clinfo returns 0 platforms. Using  4.16.0-041600rc4-generic on Ubuntu 18.04
ROCm support is still WIP according to:
https://github.com/RadeonOpenCompute/ROCm/issues/333

Not sure what other GPGPU program I could try since OpenCL is not working yet.

**Comment by Johnreidsilver** [comment]
OpenCL support for Raven Ridge is coming
https://lists.freedesktop.org/archives/amd-gfx/2018-April/021154.html

" Raven support requires some extra work that
will follow shortly, but Raven support is already included and I didn't
go out of my way to keep it out."

Now I just have to find a nice tutorial on compiling your own kernel :) rather than waiting for mainline 4.18

**Comment by Yougmark** [comment]
@Johnreidsilver That's the AMDKFD support for Raven, right?  Although I noticed updates for raven in OpenCL runtime repository, e.g., https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime/commit/3b1829cbdc2c711c7a975339c721e961baf37397#diff-0786b7bfb7cc2b95c9fc84d6b3f3534c.

Please let me know how this works out as you try!

**Comment by Johnreidsilver** [comment]
It seems we are out of luck at least until ROCm 1.9

> There is work to get Raven running it just not in ROCm 1.8. The team has been working on it internally but we are working on a number of issues.

https://github.com/RadeonOpenCompute/ROCm/issues/399

**Comment by Yougmark** [comment]
https://lists.freedesktop.org/archives/amd-gfx/2018-July/024094.html

It looks like the support for Raven is there.  Haven't tested it yet but will do.

**Comment by kentrussell** [comment]
Can you try the 1.9 release and see if the issue persists? The components should be dropping today and tomorrow.

**Comment by Johnreidsilver** [comment]
Trying Ubuntu 18.04 with latest available kernel (http://kernel.ubuntu.com) 4.19.0-041900rc4-generic on R5 2400G

 dmesg | grep kfd

`[    1.427685] kfd kfd: Initialized module
[    1.670688] kfd kfd: Allocated 3969056 bytes on gart
[    1.670870] kfd kfd: added device 1002:15dd
`

dmesg | grep amd

'[    0.000000]   AMD AuthenticAMD
[    0.000000] RAMDISK: [mem 0x318c1000-0x34c57fff]
[    0.000000] ACPI: SSDT 0x0000000092279140 001A41 (v01 AMD    AmdTable 00000001 INTL 20120913)
[    0.000000] ACPI: SSDT 0x000000009226DCE0 005367 (v02 AMD    AmdTable 00000002 MSFT 02000002)
[    0.000000] ACPI: SSDT 0x0000000092273048 00119C (v01 AMD    AMD CPU  00000001 AMD  00000001)
[    0.000000] ACPI: CRAT 0x00000000922741E8 000810 (v01 AMD    AMD CRAT 00000001 AMD  00000001)
[    0.000000] ACPI: CDIT 0x00000000922749F8 000029 (v01 AMD    AMD CDIT 00000001 AMD  00000001)
[    0.000000] ACPI: SSDT 0x0000000092274A28 002DA8 (v01 AMD    AMD AOD  00000001 INTL 20120913)
[    0.000000] ACPI: IVRS 0x0000000092277890 0000D0 (v02 AMD    AMD IVRS 00000001 AMD  00000000)
[    0.000000] ACPI: SSDT 0x0000000092277960 0000F8 (v01 AMD    AMD PT   00001000 INTL 20120913)
[    0.000000] ACPI: SSDT 0x0000000092277A58 000E96 (v01 AMD    AmdTable 00000001 INTL 20120913)
[    0.000000] ACPI: SSDT 0x00000000922788F0 000850 (v01 AMD    AmdTable 00000001 INTL 20120913)
[    0.003822] Spectre V2 : Mitigation: Full AMD retpoline
[    0.016008] smpboot: CPU0: AMD Ryzen 5 2400G with Radeon Vega Graphics (family: 0x17, model: 0x11, stepping: 0x0)
[    0.016008] Performance Events: Fam17h core perfctr, AMD PMU driver.
[    0.646911] AMD-Vi: Unable to write to IOMMU perf counter.
[    0.650162] AMD-Vi: Found IOMMU at 0000:00:00.2 cap 0x40
[    0.650162] AMD-Vi: Extended features (0x4f77ef22294ada):
[    0.650165] AMD-Vi: Interrupt remapping enabled
[    0.650165] AMD-Vi: virtual APIC enabled
[    0.744597] AMD-Vi: Lazy IO/TLB flushing enabled
[    0.745311] amd_uncore: AMD NB counters detected
[    0.745321] amd_uncore: AMD LLC counters detected
[    1.425009] AMD IOMMUv2 driver by Joerg Roedel <jroedel@suse.de>
'

'[    1.422955] [drm] amdgpu kernel modesetting enabled.
[    1.425009] AMD IOMMUv2 driver by Joerg Roedel <jroedel@suse.de>
[    1.427636] Parsing CRAT table with 1 nodes
[    1.427637] Creating topology SYSFS entries
[    1.427659] Topology: Add APU node [0x0:0x0]
[    1.427660] Finished initializing topology
[    1.427685] kfd kfd: Initialized module
[    1.427791] checking generic (e0000000 3f0000) vs hw (e0000000 10000000)
[    1.427791] fb: switching to amdgpudrmfb from VESA VGA
[    1.427811] Console: switching to colour dummy device 80x25
[    1.428070] [drm] initializing kernel modesetting (RAVEN 0x1002:0x15DD 0x1043:0x876B 0xC6).
[    1.428079] [drm] register mmio base: 0xFE600000
[    1.428079] [drm] register mmio size: 524288
[    1.428101] [drm] add ip block number 0 <soc15_common>
[    1.428102] [drm] add ip block number 1 <gmc_v9_0>
[    1.428103] [drm] add ip block number 2 <vega10_ih>
[    1.428103] [drm] add ip block number 3 <psp>
[    1.428104] [drm] add ip block number 4 <powerplay>
[    1.428104] [drm] add ip block number 5 <dm>
[    1.428105] [drm] add ip block number 6 <gfx_v9_0>
[    1.428106] [drm] add ip block number 7 <sdma_v4_0>
[    1.428106] [drm] add ip block number 8 <vcn_v1_0>
[    1.428132] [drm] VCN decode is enabled in VM mode
[    1.428132] [drm] VCN encode is enabled in VM mode
[    1.428132] [drm] VCN jpeg decode is enabled in VM mode
[    1.452482] [drm] BIOS signature incorrect 0 0
[    1.452509] ATOM BIOS: 113-RAVEN-106
[    1.452548] [drm] vm size is 262144 GB, 4 levels, block size is 9-bit, fragment size is 9-bit
[    1.452558] amdgpu 0000:07:00.0: VRAM: 1024M 0x000000F400000000 - 0x000000F43FFFFFFF (1024M used)
[    1.452559] amdgpu 0000:07:00.0: GART: 1024M 0x000000F500000000 - 0x000000F53FFFFFFF
[    1.452562] [drm] Detected VRAM RAM=1024M, BAR=1024M
[    1.452562] [drm] RAM width 64bits DDR4
[    1.452640] [TTM] Zone  kernel: Available graphics memory: 3564326 kiB
[    1.452641] [TTM] Zone   dma32: Available graphics memory: 2097152 kiB
[    1.452641] [TTM] Initializing pool allocator
[    1.452644] [TTM] Initializing DMA pool allocator
[    1.452677] [drm] amdgpu: 1024M of VRAM memory ready
[    1.452677] [drm] amdgpu: 3072M of GTT memory ready.
[    1.452683] [drm] GART: num cpu pages 262144, num gpu pages 262144
[    1.452849] [drm] PCIE GART of 1024M enabled (table at 0x000000F400900000).
[    1.453930] [drm] use_doorbell being set to: [true]
[    1.454058] [drm] Found VCN firmware Version: 1.73 Family ID: 18
[    1.454060] [drm] PSP loading VCN firmware
[    1.560023] usb 1-8: new low-speed USB device number 3 using xhci_hcd
[    1.617263] amdgpu: [powerplay] dpm has been enabled
[    1.617336] [drm] DM_PPLIB: values for Invalid clock
[    1.617337] [drm] DM_PPLIB:   400000 in kHz
[    1.617337] [drm] DM_PPLIB:   933000 in kHz
[    1.617338] [drm] DM_PPLIB:   1067000 in kHz
[    1.617338] [drm] DM_PPLIB:   1200000 in kHz
[    1.617339] [drm] DM_PPLIB: values for Invalid clock
[    1.617339] [drm] DM_PPLIB:   300000 in kHz
[    1.617339] [drm] DM_PPLIB:   600000 in kHz
[    1.617340] [drm] DM_PPLIB:   626000 in kHz
[    1.617340] [drm] DM_PPLIB:   654000 in kHz
[    1.620222] [drm] Display Core initialized with v3.1.59!
[    1.646391] [drm] Supports vblank timestamp caching Rev 2 (21.10.2013).
[    1.646391] [drm] Driver supports precise vblank timestamp query.
[    1.669685] [drm] VCN decode and encode initialized successfully.
[    1.670688] kfd kfd: Allocated 3969056 bytes on gart
[    1.670709] Topology: Add APU node [0x15dd:0x1002]
[    1.670870] kfd kfd: added device 1002:15dd
[    1.672298] [drm] fb mappable at 0xA1100000
[    1.672299] [drm] vram apper at 0xA0000000
[    1.672299] [drm] size 4096000
[    1.672300] [drm] fb depth is 24
[    1.672300] [drm]    pitch is 5120
[    1.672353] fbcon: amdgpudrmfb (fb0) is primary device
[    1.708276] Console: switching to colour frame buffer device 160x50
[    1.716523] ata9: SATA link down (SStatus 0 SControl 300)
[    1.726873] amdgpu 0000:07:00.0: fb0: amdgpudrmfb frame buffer device
[    1.740090] amdgpu 0000:07:00.0: ring 0(gfx) uses VM inv eng 4 on hub 0
[    1.740091] amdgpu 0000:07:00.0: ring 1(comp_1.0.0) uses VM inv eng 5 on hub 0
[    1.740093] amdgpu 0000:07:00.0: ring 2(comp_1.1.0) uses VM inv eng 6 on hub 0
[    1.740094] amdgpu 0000:07:00.0: ring 3(comp_1.2.0) uses VM inv eng 7 on hub 0
[    1.740095] amdgpu 0000:07:00.0: ring 4(comp_1.3.0) uses VM inv eng 8 on hub 0
[    1.740096] amdgpu 0000:07:00.0: ring 5(comp_1.0.1) uses VM inv eng 9 on hub 0
[    1.740097] amdgpu 0000:07:00.0: ring 6(comp_1.1.1) uses VM inv eng 10 on hub 0
[    1.740098] amdgpu 0000:07:00.0: ring 7(comp_1.2.1) uses VM inv eng 11 on hub 0
[    1.740099] amdgpu 0000:07:00.0: ring 8(comp_1.3.1) uses VM inv eng 12 on hub 0
[    1.740101] amdgpu 0000:07:00.0: ring 9(kiq_2.1.0) uses VM inv eng 13 on hub 0
[    1.740102] amdgpu 0000:07:00.0: ring 10(sdma0) uses VM inv eng 4 on hub 1
[    1.740103] amdgpu 0000:07:00.0: ring 11(vcn_dec) uses VM inv eng 5 on hub 1
[    1.740104] amdgpu 0000:07:00.0: ring 12(vcn_enc0) uses VM inv eng 6 on hub 1
[    1.740105] amdgpu 0000:07:00.0: ring 13(vcn_enc1) uses VM inv eng 7 on hub 1
[    1.740106] amdgpu 0000:07:00.0: ring 14(vcn_jpeg) uses VM inv eng 8 on hub 1
[    1.742714] [drm] Initialized amdgpu 3.27.0 20150101 for 0000:07:00.0 on minor 0
'

**Comment by kentrussell** [comment]
If it's working, I'll close the issue. Glad we got this resolved!
